### PR TITLE
Msg Difference Per Period

### PR DESCRIPTION
**Pull request by DanielMwendwa**
…periods (the time period for firebase is a constant number)

### CODE REVIEW

File: mno_analysis_tools/compute_msg_difference_btwn_periods.py [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+    parser.add_argument("target_operator", metavar="target-operator",
+                        help="Operator to compute message difference between two firebase time periods")
+    parser.add_argument("target_message_direction", metavar="target-message-direction", choices=('in', 'out'),
+                        help="Direction of messages to limit the search for downtime to")
+    parser.add_argument("start_date", metavar="start-date", type=lambda s: isoparse(s),
+                        help="The start date as ISO 8601 string from which the number of messages will be computed")
+    parser.add_argument("end_date", metavar="end-date", type=lambda s: isoparse(s),
+                        help="The end date as ISO 8601 string to which the number of messages computation will end")
+    parser.add_argument("-t", "--time_frame", action="store_true", metavar="time-frame", type=lambda s: datetime.strptime(s, '%H:%M:%S'),
+                        default="00:10:00", help="The time frame (HH:MM:SS) to generate dates in intervals between the start and end date")
```
Comment (IsaackMwenda) [inline mno_analysis_tools/compute_msg_difference_btwn_periods.py]: I think the default should be "00:00:10" ?

File: mno_analysis_tools/compute_msg_difference_btwn_periods.py [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+        time_frame = args.time_frame
+
+    with open(raw_messages_input_file_path, mode="r") as f:
+        log.info("Loading messages from {raw_messages_input_file_path}...")
+        input = json.load(f)
+        messages = [Message.deserialize(val) for val in input]
+        log.info(f"Loaded {len(messages)} messages")
+
+    # Filter messages based on the target operator and target direction of the message
+    log.info(f"Filtering messages based on operator {target_operator} and "
```
Comment (IsaackMwenda) [inline mno_analysis_tools/compute_msg_difference_btwn_periods.py]:
```suggestion
    log.info(f"Filtering messages based on {target_operator} and "
```

File: mno_analysis_tools/compute_msg_difference_btwn_periods.py [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+             "message direction {target_message_direction} from {len(messages)} total messages ")
```
Comment (IsaackMwenda) [inline mno_analysis_tools/compute_msg_difference_btwn_periods.py]:
```suggestion
             "{target_message_direction} message direction from {len(messages)} total messages ")
```

File: mno_analysis_tools/compute_msg_difference_btwn_periods.py [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+    for msg in messages:
+        if msg.urn.startswith("tel:"):
+            operator = PhoneCleaner.clean_operator(msg.urn.split(":")[1])
+        else:
+            operator = msg.urn.split(":")[0]
+        if operator == target_operator and msg.direction == target_message_direction:
+            msg_direction = msg.direction
+            filtered_messages.append(msg)
+    log.info(
+        f" returning {len(filtered_messages)} {target_operator} {target_message_direction} messages")
```
Comment (IsaackMwenda) [inline mno_analysis_tools/compute_msg_difference_btwn_periods.py]: read the logs and make sure they are coherent, reformat where necessary.

File: mno_analysis_tools/compute_msg_difference_btwn_periods.py [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+
+    time_interval = timedelta(hours=time_frame.hour,
+                              minutes=time_frame.minute, seconds=time_frame.second)
+
+    date_time_bounds = date_time_range(start_date, end_date, time_interval)
+
+    # Compute number of messages between two datetime bounds i.e `PreviousMessageTimestamp` and
```
Comment (IsaackMwenda) [inline mno_analysis_tools/compute_msg_difference_btwn_periods.py]:
```suggestion
    # Compute the number of messages between two firebase time bounds i.e `PreviousMessageTimestamp` and
```

File: mno_analysis_tools/compute_msg_difference_btwn_periods.py [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+    # `NextMessageTimestamp` to get number of mesages per period and relate each quantity
```
Comment (IsaackMwenda) [inline mno_analysis_tools/compute_msg_difference_btwn_periods.py]: format this to match the new logic

File: mno_analysis_tools/compute_msg_difference_btwn_periods.py [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+    current = start
+    intervals = []
+    while current < end:
+        intervals.append(current)
+        current += delta
+    return intervals
+
+
+log = Logger(__name__)
+log.set_project_name("ComputeMessagesBetweenTwoFirebaseTimePeriods")
```
Comment (IsaackMwenda) [inline mno_analysis_tools/compute_msg_difference_btwn_periods.py]: name the file to match this

File: mno_analysis_tools/compute_msg_difference_btwn_periods.py [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+            "Operator": operator,
+            "MessageDirection": msg_direction,
+            "PeriodStart": messages_per_period[index]["PeriodStart"],
+            "PeriodBetween": messages_per_period[index]["PeriodEnd"],
+            "PeriodEnd": messages_per_period[next_index]["PeriodEnd"],
+            "MessageDifference": messages_per_period[next_index]["NumberOfMessages"] - messages_per_period[index]["NumberOfMessages"]
+        })
+
+    log.info(
+        f"writing messages_per_period and  message_difference_per_period json files...")
```
Comment (IsaackMwenda) [inline mno_analysis_tools/compute_msg_difference_btwn_periods.py]: we are not writing both files here

File: mno_analysis_tools/compute_msg_difference_btwn_periods.py [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+    with open(message_difference_output_file_path, mode="w") as f:
+        json.dump(messages_per_period, f)
```
Comment (IsaackMwenda) [inline mno_analysis_tools/compute_msg_difference_btwn_periods.py]: I think we you want messages_difference_per_two_firebase_period_output_file_path

File: mno_analysis_tools/compute_msg_difference_btwn_periods.py (hunk 8 above)
Comment (DanielMwendwa) [reply mno_analysis_tools/compute_msg_difference_btwn_periods.py]: yeah

File: mno_analysis_tools/compute_msg_difference_btwn_two_firebase_time_periods.py [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+            "Operator": operator,
+            "MessageDirection": msg_direction,
+            "PeriodStart": messages_per_period[index]["PeriodStart"],
+            "PeriodBetween": messages_per_period[index]["PeriodEnd"],
+            "PeriodEnd": messages_per_period[next_index]["PeriodEnd"],
+            "MessageDifference": messages_per_period[next_index]["NumberOfMessages"] - messages_per_period[index]["NumberOfMessages"]
+        })
+
+    log.info(f"writing message_difference_per_period json file...")
+    with open(messages_difference_per_two_firebase_period_output_file_path, mode="w") as f:
```
Comment (IsaackMwenda) [inline mno_analysis_tools/compute_msg_difference_btwn_two_firebase_time_periods.py]:
```suggestion
    with open(messages_difference_per_two_firebase_time_period_output_file_path, mode="w") as f:
```

**Pull request merged**